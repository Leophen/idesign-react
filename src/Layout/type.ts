export interface UsualProps {
  /**
   * 组件自定义类名
   */
  className?: string;
  /**
   * 内容
   */
  children?: React.ReactNode;
  /**
   * 组件自定义样式
   */
  style?: React.CSSProperties;
}

export interface LayoutProps extends UsualProps {}

export interface LayoutHeaderProps extends UsualProps {
  /**
   * 顶栏高度。样式表（class）中定义的默认高度为：64px
   * @default ''
   */
  height?: React.CSSProperties["height"];
}

export interface LayoutAsideProps extends UsualProps {
  /**
   * 侧边栏宽度。样式表（class）中定义的默认宽度为：232px
   * @default ''
   */
  width?: React.CSSProperties["width"];
}

export interface LayoutContentProps extends UsualProps {}

export interface LayoutFooterProps extends UsualProps {
  /**
   * 底栏高度。样式表（class）中定义的默认高度为：24px
   * @default ''
   */
  height?: React.CSSProperties["height"];
}
