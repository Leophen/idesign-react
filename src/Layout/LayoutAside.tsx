import React from 'react';
import classNames from 'classnames';
import './index.scss';
import { LayoutAsideProps } from './type';

const LayoutAside: React.FC<LayoutAsideProps> = (props) => {
  const { width = 240, className, style, children, ...otherAsideProps } = props;

  const asideClassNames = classNames('i-layout--aside', className);
  const asideWidth = typeof width === 'number' ? `${width}px` : String(width);

  const asideStyle = {
    width: asideWidth,
    maxWidth: asideWidth,
    minWidth: asideWidth,
    flex: `0 0 ${asideWidth}`,
    ...style,
  };

  return (
    <aside className={asideClassNames} style={asideStyle} {...otherAsideProps}>
      {children}
    </aside>
  );
};

export default LayoutAside;
