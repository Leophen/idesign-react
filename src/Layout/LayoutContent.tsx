import React from 'react';
import classNames from 'classnames';
import './index.scss';
import { LayoutContentProps } from './type';

const LayoutContent: React.FC<LayoutContentProps> = (props) => {
  const { children, className, style, ...restProps } = props;
  return (
    <main className={classNames('i-layout--content', className)} style={style} {...restProps}>
      {children}
    </main>
  );
};

export default LayoutContent;
