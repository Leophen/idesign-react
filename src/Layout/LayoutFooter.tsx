import React from 'react';
import classNames from 'classnames';
import './index.scss';
import { LayoutFooterProps } from './type';

const LayoutFooter: React.FC<LayoutFooterProps> = (props) => {
  const { children, className, style, ...restProps } = props;
  return (
    <footer className={classNames('i-layout--footer', className)} style={style} {...restProps}>
      {children}
    </footer>
  );
};

export default LayoutFooter;
