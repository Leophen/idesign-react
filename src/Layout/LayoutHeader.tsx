import React from 'react';
import classNames from 'classnames';
import './index.scss';
import { LayoutHeaderProps } from './type';

const LayoutHeader: React.FC<LayoutHeaderProps> = (props) => {
  const { children, className, style, ...restProps } = props;
  return (
    <header className={classNames('i-layout--header', className)} style={style} {...restProps}>
      {children}
    </header>
  );
};

export default LayoutHeader;
