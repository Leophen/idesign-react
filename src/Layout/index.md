---
nav:
  title: 组件
  path: /components
group:
  title: 基础组件
  order: 1
---

# Layout 布局

用于组织网页的框架结构。

## 基本用法

`<Layout>` 容器、可包裹子组件 `<Header>`、`<Footer>`、`<Aside>`、`<Content>`。

```tsx
import React from 'react';
import { Layout } from 'idesign-react';

const { Header, Content, Footer, Aside } = Layout;

const App = () => {
  return (
    <Layout>
      <Header>Header</Header>
      <Content>Content</Content>
      <Footer>Copyright @ 2019-2021 iDesign. All Rights Reserved</Footer>
    </Layout>
  );
};

export default App;
```

## 侧边导航布局

当 `<Layout>` 容器子元素中包含 `<Aside>` 侧边栏组件时，`<Layout>` 容器的全部子元素会水平排列，否则会垂直排列。

```tsx
import React from 'react';
import { Layout } from 'idesign-react';

const { Header, Content, Footer, Aside } = Layout;

const App = () => {
  return (
    <Layout>
      <Aside>Aside</Aside>
      <Layout>
        <Content>Content</Content>
        <Footer>Copyright @ 2019-2021 iDesign. All Rights Reserved</Footer>
      </Layout>
    </Layout>
  );
};

export default App;
```

## 组合导航布局

可根据所需场景自由组合。

```tsx
import React from 'react';
import { Layout } from 'idesign-react';

const { Header, Content, Footer, Aside } = Layout;

const App = () => {
  return (
    <>
      <Layout>
        <Header>Header</Header>
        <Layout>
          <Aside>Aside</Aside>
          <Layout>
            <Content>Content</Content>
            <Footer>Copyright @ 2019-2021 iDesign. All Rights Reserved</Footer>
          </Layout>
        </Layout>
      </Layout>

      <br />
      <br />

      <Layout>
        <Header>Header</Header>
        <Layout>
          <Layout>
            <Content>Content</Content>
            <Footer>Copyright @ 2019-2021 iDesign. All Rights Reserved</Footer>
          </Layout>
          <Aside>Aside</Aside>
        </Layout>
      </Layout>
    </>
  );
};

export default App;
```

## 自定义侧边栏

`Aside` 提供 `width` 属性来自定义宽度。

```tsx
import React from 'react';
import { Layout } from 'idesign-react';

const { Header, Content, Footer, Aside } = Layout;

const App = () => {
  return (
    <>
      <Layout>
        <Header>Header</Header>
        <Layout>
          <Aside width={100}>Aside</Aside>
          <Layout>
            <Content>Content</Content>
            <Footer>Copyright @ 2019-2021 iDesign. All Rights Reserved</Footer>
          </Layout>
        </Layout>
      </Layout>

      <br />
      <br />

      <Layout>
        <Header>Header</Header>
        <Layout>
          <Aside width="100px">Aside</Aside>
          <Layout>
            <Content>Content</Content>
            <Footer>Copyright @ 2019-2021 iDesign. All Rights Reserved</Footer>
          </Layout>
        </Layout>
      </Layout>
    </>
  );
};

export default App;
```

## Aside API

| 属性      | 说明       | 类型            | 默认值 |
| --------- | ---------- | --------------- | ------ |
| children  | 内容       | `ReactNode`     | `--`   |
| className | 自定义类名 | `string`        | `--`   |
| style     | 自定义样式 | `CSSProperties` | `--`   |
| width     | 宽度       | `string`        | `240`  |
