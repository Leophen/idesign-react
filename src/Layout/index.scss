.i-layout {
  display: flex;
  flex: auto;
  flex-direction: column;
  &.i-layout-has-aside {
    flex-direction: row;
  }
}

.i-layout--header {
  box-sizing: border-box;
}

.i-layout--aside {
  box-sizing: border-box;
}

.i-layout--content {
  box-sizing: border-box;
}

.i-layout--footer {
  box-sizing: border-box;
}
