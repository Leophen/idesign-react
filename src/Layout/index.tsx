import React, { useEffect, useState } from 'react';
import classNames from 'classnames';
import './index.scss';

export interface UsualProps {
  /**
   * 组件自定义类名
   */
  className?: string;
  /**
   * 内容
   */
  children?: React.ReactNode;
  /**
   * 组件自定义样式
   */
  style?: React.CSSProperties;
}

export interface LayoutProps extends UsualProps {}

export interface HeaderProps extends UsualProps {
  /**
   * 顶栏高度。样式表（class）中定义的默认高度为：64px
   * @default ''
   */
  height?: React.CSSProperties["height"];
}

export interface AsideProps extends UsualProps {
  /**
   * 侧边栏宽度。样式表（class）中定义的默认宽度为：232px
   * @default ''
   */
  width?: React.CSSProperties["width"];
}

export interface ContentProps extends UsualProps {}

export interface FooterProps extends UsualProps {
  /**
   * 底栏高度。样式表（class）中定义的默认高度为：24px
   * @default ''
   */
  height?: React.CSSProperties["height"];
}

const Header: React.FC<HeaderProps> = (props) => {
  const { children, className, style, ...restProps } = props;
  return (
    <header className={classNames('i-layout--header', className)} style={style} {...restProps}>
      {children}
    </header>
  );
};

const Aside: React.FC<AsideProps> = (props) => {
  const { width = 240, className, style, children, ...otherAsideProps } = props;

  const asideClassNames = classNames('i-layout--aside', className);
  const asideWidth = typeof width === 'number' ? `${width}px` : String(width);

  const asideStyle = {
    width: asideWidth,
    maxWidth: asideWidth,
    minWidth: asideWidth,
    flex: `0 0 ${asideWidth}`,
    ...style,
  };

  return (
    <aside className={asideClassNames} style={asideStyle} {...otherAsideProps}>
      {children}
    </aside>
  );
};

const Content: React.FC<ContentProps> = (props) => {
  const { children, className, style, ...restProps } = props;
  return (
    <main className={classNames('i-layout--content', className)} style={style} {...restProps}>
      {children}
    </main>
  );
};

const Footer: React.FC<FooterProps> = (props) => {
  const { children, className, style, ...restProps } = props;
  return (
    <footer className={classNames('i-layout--footer', className)} style={style} {...restProps}>
      {children}
    </footer>
  );
};

/**
 * 布局组件
 */
const Layout: React.FC<LayoutProps> & {
  Header: React.ElementType;
  Aside: React.ElementType;
  Content: React.ElementType;
  Footer: React.ElementType;
} = (props) => {
  const { children, className, style, ...restProps } = props;

  const [asides, setAsides] = useState([]);

  useEffect(() => {
    React.Children.forEach(children, (child: any) => {
      if (child.type === Aside) setAsides([child] as any);
    });
  }, [children]);

  const layoutClassNames = classNames(
    'i-layout',
    !!asides.length && 'i-layout-has-aside',
    className,
  );

  return (
    <div className={layoutClassNames} style={style} {...restProps}>
      {children}
    </div>
  );
};

Layout.Header = Header;
Layout.Aside = Aside;
Layout.Content = Content;
Layout.Footer = Footer;

Header.displayName = 'Header';
Aside.displayName = 'Aside';
Content.displayName = 'Content';
Footer.displayName = 'Footer';

Layout.displayName = 'Layout';

export default Layout;
