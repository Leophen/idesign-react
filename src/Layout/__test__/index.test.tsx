import { render } from '@testing-library/react';
import Layout from '../index';

const { Header, Content, Footer, Aside } = Layout;

describe('Layout 组件测试', () => {
  it('create', () => {
    const { asFragment } = render(
      <Layout>
        <Header>Header</Header>
        <Layout>
          <Aside>Aside</Aside>
          <Layout>
            <Content>Content</Content>
            <Footer>Copyright @ 2019-2021 iDesign. All Rights Reserved</Footer>
          </Layout>
        </Layout>
      </Layout>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('width', () => {
    const wrapper = render(
      <Layout>
        <Header>Header</Header>
        <Layout>
          <Aside data-testid="aside" width={100}>Aside</Aside>
          <Layout>
            <Content>Content</Content>
            <Footer>Copyright @ 2019-2021 iDesign. All Rights Reserved</Footer>
          </Layout>
        </Layout>
      </Layout>
    );
    expect(wrapper.getByTestId('aside')).toHaveStyle('width: 100px');
  })
});

