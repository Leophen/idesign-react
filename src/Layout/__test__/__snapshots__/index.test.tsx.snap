// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout 组件测试 create 1`] = `
<DocumentFragment>
  <div
    class="i-layout"
  >
    <header
      class="i-layout--header"
    >
      Header
    </header>
    <div
      class="i-layout i-layout-has-aside"
    >
      <aside
        class="i-layout--aside"
        style="width: 240px; max-width: 240px; min-width: 240px; flex: 0 0 240px;"
      >
        Aside
      </aside>
      <div
        class="i-layout"
      >
        <main
          class="i-layout--content"
        >
          Content
        </main>
        <footer
          class="i-layout--footer"
        >
          Copyright @ 2019-2021 iDesign. All Rights Reserved
        </footer>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
